src/store/reducers/favorites.js

```diff
@@ -1,33 +1,31 @@
-const initialState = {
-  songs: [],
+import { Map, List } from 'immutable';
+
+import {
+  INITIALIZE_FAVORITES, ADD_FAVORITE, REMOVE_FAVORITE, TOGGLE_IS_FAVORITE_PLAYING,
+} from '../constants';
+
+const initialState = Map({
+  songs: List(),
   isPlaying: false,
-};
+});
 
 const favoritesReducer = (state = initialState, { type, payload }) => {
   switch (type) {
-  case 'SET_FAVORITES':
-    return {
-      ...state,
-      songs: payload,
-    };
-
-  case 'ADD_FAVORITE':
-    return {
-      ...state,
-      songs: state.songs.concat(payload),
-    };
-
-  case 'REMOVE_FAVORITE':
-    return {
-      ...state,
-      songs: state.songs.filter((song) => song.song_name !== payload.song_name),
-    };
-
-  case 'TOGGLE_IS_PLAYING':
-    return {
-      ...state,
-      isPlaying: payload,
-    };
+  case INITIALIZE_FAVORITES:
+    return state.set('songs', payload);
+
+  case ADD_FAVORITE: {
+    const favoritesUpdated = state.get('songs').push(payload);
+    return state.set('songs', favoritesUpdated);
+  }
+
+  case REMOVE_FAVORITE: {
+    const favoritesUpdated = state.songs.filter((song) => song.song_name !== payload.song_name);
+    return state.set('songs', favoritesUpdated);
```

Comment:
Let's change this to:

```
const itemIndex = state.get('songs').findIndex(song => song.get('song_name') === payload.get('song_name'));
return itemIndex > -1 ? state.deleteIn(['songs', itemIndex]) : state;
```

In this way we won't mutate the state if the index for some reason is not found. Also, you should use `get()` to access Map's props.